lib/rack/handler/puma.rb

```diff
@@ -1,114 +1,136 @@
-          puts "* Gracefully stopping, waiting for requests to finish"
-          launcher.stop
-          puts "* Goodbye!"
+        if default_options[:Host]
+          file_config.set_default_host(default_options[:Host])
         end
+        self.set_host_port_to_config(default_options[:Host], default_options[:Port], default_config)
+
+        user_config.app app
       end
+      conf
+    end
+
+    def run(app, **options)
+      conf   = self.config(app, options)
+
+      log_writer = options.delete(:Silent) ? ::Puma::LogWriter.strings : ::Puma::LogWriter.stdio
+
+      launcher = ::Puma::Launcher.new(conf, :log_writer => log_writer)
 
-      def self.valid_options
-        {
-          "Host=HOST"       => "Hostname to listen on (default: localhost)",
-          "Port=PORT"       => "Port to listen on (default: 8080)",
-          "Threads=MIN:MAX" => "min:max threads to use (default 0:16)",
-          "Verbose"         => "Don't report each request (default: false)"
-        }
+      yield launcher if block_given?
+      begin
+        launcher.run
+      rescue Interrupt
+        puts "* Gracefully stopping, waiting for requests to finish"
+        launcher.stop
+        puts "* Goodbye!"
       end
+    end
 
-      def self.set_host_port_to_config(host, port, config)
-        config.clear_binds! if host || port
-
-        if host && (host[0,1] == '.' || host[0,1] == '/')
-          config.bind "unix://#{host}"
-        elsif host && host =~ /^ssl:\/\//
-          uri = URI.parse(host)
-          uri.port ||= port || ::Puma::Configuration::DEFAULTS[:tcp_port]
-          config.bind uri.to_s
-        else
-
-          if host
-            port ||= ::Puma::Configuration::DEFAULTS[:tcp_port]
-          end
-
-          if port
-            host ||= ::Puma::Configuration::DEFAULTS[:tcp_host]
-            config.port port, host
-          end
+    def valid_options
+      {
+        "Host=HOST"       => "Hostname to listen on (default: localhost)",
+        "Port=PORT"       => "Port to listen on (default: 8080)",
+        "Threads=MIN:MAX" => "min:max threads to use (default 0:16)",
+        "Verbose"         => "Don't report each request (default: false)"
+      }
+    end
+
+    def set_host_port_to_config(host, port, config)
+      config.clear_binds! if host || port
+
+      if host && (host[0,1] == '.' || host[0,1] == '/')
+        config.bind "unix://#{host}"
+      elsif host && host =~ /^ssl:\/\//
+        uri = URI.parse(host)
+        uri.port ||= port || ::Puma::Configuration::DEFAULTS[:tcp_port]
+        config.bind uri.to_s
+      else
+
+        if host
+          port ||= ::Puma::Configuration::DEFAULTS[:tcp_port]
+        end
+
+        if port
+          host ||= ::Puma::Configuration::DEFAULTS[:tcp_host]
+          config.port port, host
         end
       end
     end
+  end
+end
 
-    register :puma, Puma
+# rackup was removed in Rack 3, it is now a separate gem
+if Object.const_defined? :Rackup
```

Comment:
Can we just combine this with what's in `rack_default`? 🤔

Reply:
No.  As mentioned above, Rackup currently will look for several servers, and Puma is the first one.  It determines if Puma is available by requiring `rack/handler/puma`.

For info on `rack_default`, see https://github.com/puma/puma/pull/194.